compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -756,6 +756,46 @@ impl<'a> Parser<'a> {
             }
         }
 
+        // We could've continued to parse bounds but we found a type instead of a trait.
+        if !self.can_begin_bound() && self.check_type() {
+            let ty = self.parse_ty().unwrap();
```

Comment:
This can fail: `check_type` only checks if the current token can *begin* a type not that the upcoming type is valid. E.g. `T: &fn<>()` would crash the compiler in your current patch. Either bubble up the error with `?` or preferably (?) pattern match on the `Result` and only show the custom suggestion if `Ok(ty)`.

Reply:
Hmm, that example doesn't panic and I'm not sure why. In any case, I've made the change.

Reply:
Ah, yeah my bad. `fn<>()` itself gets recovered. I guess the bound `T: &fn<const>()` would've actually panicked.